packages/runtime/src/state/react-builder-preview.ts

```diff
@@ -648,7 +661,70 @@ function makeswiftApiClientSyncMiddleware(
   }
 }
 
-export type Store = ReduxStore<State, Action> & { dispatch: Dispatch }
+class MessageChannel {
+  private channel: MessagePort | null = null
+  private bufferedMessages: [Action, Transferable[]?][] = []
+
+  public postMessage(message: any, transferables?: Transferable[]) {
+    if (this.channel) {
+      this.channel.postMessage(message, transferables ?? [])
+    } else {
+      this.bufferedMessages.push([message, transferables])
+    }
+  }
+
+  public setup(onMessage: (event: MessageEvent<Action>) => void) {
+    const channel = new window.MessageChannel()
+    channel.port1.onmessage = onMessage
+
+    // connect channel to the parent window, see
+    // https://developer.mozilla.org/en-US/docs/Web/API/Channel_Messaging_API
+    window.parent.postMessage(channel.port2, '*', [channel.port2])
+
+    this.channel = channel.port1
+  }
+
+  public dispatchBuffered() {
+    console.assert(this.channel != null, 'channel is not setup')
+
+    this.bufferedMessages.forEach(([message, transferables]) => {
+      this.postMessage(message, transferables ?? [])
+    })
```

Comment:
Are there any potential issues here with messages pushing more messages resulting in a potential loop. Does `forEach` take a snapshot of the array or should be first get a clone of the buffered messages before we loop?

Reply:
Good point about making sure we don't loop if the precondition fails. I think this should be just `this.channel?.postMessage`.

Reply:
Well, yes, but also what I mean is something like:

```
const bufferedMessages = this.bufferedMessages.slice()

bufferedMessages.forEach(// ...
```

This makes it so that it `this.bufferedMessages` is mutated in the loop, we don't go back in. This might be unnecessary depending on how `forEach` works. If we were using a `for` loop on the mutable array, for example, it would definitely be an issue.

Reply:
That's how I understood you, but I felt like changing the loop body to `this.channel?.postMessage` eliminates the possibility of `bufferedMessages` being mutated. If `channel` is not null and we actually do post a message that generates a response from the builder and dispatches another message, because `channel` is not null, that message won't be buffered and will be just sent to the builder. 

We could do `.slice()` just to "future-proof" this, but I feel like that would actually make me think that this "re-buffering" scenario is possible (since we don't `slice()` everywhere as a general rule), which would be confusing. Does this make sense?

Reply:
Oh, whoops, I totally missed that you were using the channel directly 🤦🏻‍♂️  This makes total sense.